sklearn/preprocessing/data.py

```diff
@@ -1538,10 +1540,45 @@ def transform(self, X):
                         columns.append(bias)
                 XP = sparse.hstack(columns, dtype=X.dtype).tocsc()
             else:
+                # When the matrix is small, it is faster to transpose
+                # the matrix first to multiply contiguous memory segments.
+                n = X.shape[1]
+
                 XP = np.empty((n_samples, self.n_output_features_),
                               dtype=X.dtype, order=self.order)
-                for i, comb in enumerate(combinations):
-                    XP[:, i] = X[:, comb].prod(1)
+
+                if self.include_bias:
+                    XP[:, 0] = 1
+
+                pos = 1 if self.include_bias else 0
+                for d in range(0, self.degree):
+                    if d == 0:
+                        XP[:, pos:pos + n] = X
+                        index = list(range(pos, pos + n))
+                        pos += n
+                        index.append(pos)
+                    else:
+                        new_index = []
+                        end = index[-1]
+                        for i in range(0, n):
+                            a = index[i]
+                            new_index.append(pos)
+                            if self.interaction_only:
+                                start = a + index[i+1] - index[i]
+                            else:
+                                start = a
```

Comment:
This could be simplified as:

```python
                            new_index.append(pos)
                            start = index[i]
                            if self.interaction_only:
                                start += index[i+1] - index[i]
```

Reply:
Done